### PR TITLE
Add os/file.Truncate

### PR DESCRIPTION
**Pull request by leongross**
Add `file.Truncate()` to `file.go`

### CODE REVIEW

File: src/os/truncate_test.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+import (
+	"os"
+	. "os"
+	"path/filepath"
+	"testing"
+)
+
+func TestTruncate(t *testing.T) {
+	tmpDir, _ := os.MkdirTemp("", "TestTruncate")
```
Comment (rminnich) [inline src/os/truncate_test.go]: you can also use t.TempDir and avoid lines 17-21

**Comment by 10000TB** [comment]
any update on next steps ?

**Comment by leongross** [comment]
CI error does not seem to be related to the `Truncate` code, maybe another compiler issue? @deadprogram @aykevl, is this related to https://github.com/tinygo-org/tinygo/issues/3871?

```
no Go files in D:\a\tinygo\tinygo\tests\testing
ld.lld: error: undefined symbol: notARealFunction
>>> referenced by main
>>>               C:\Users\RUNNER~1\AppData\Local\Temp\tinygo1547056251\main.o
failed to run tool: ld.lld
--- FAIL: TestBuild (0.00s)
    --- FAIL: TestBuild/Host (0.00s)
        --- FAIL: TestBuild/Host/rand.go (36.58s)
            main_test.go:395: output did not match (expected 35 bytes, got 0 bytes):
FAIL
FAIL	github.com/tinygo-org/tinygo	110.872s
```

**Comment by 10000TB** [comment]
@leongross that seems fixed in 0.29.0

**Comment by deadprogram** [comment]
@leongross can you please rebase this PR against the latest `dev`? Thanks!

**Comment by leongross** [comment]
@deadprogram rebase done

**Comment by 10000TB** [comment]
Looking for review / approval ! TY

**Comment by leongross** [comment]
@deadprogram @aykevl

**Comment by dgryski** [comment]
Conflicts after the wasip2 merge.  This probably needs a wasip2 implementation but I'm fine to merge this with a `ErrNotImplemented` error for now.

**Comment by dgryski** [comment]
You should be able to stub this for wasip2 with this addition to `libc_waisp2.go`:
```
// int truncate(const char *path, off_t length);
//
//export truncate
func truncate(path *byte, length int64) int {
	libcErrno = ENOSYS
	return -1
}
```

**Comment by leongross** [comment]
On linux the error returned by the test is `errno = 52`, which means 'Invalid exchange'
Not sure how that would be triggered here, since the man page for truncate does not mention that error number.

**Comment by deadprogram** [comment]
@leongross please see error here https://github.com/tinygo-org/tinygo/actions/runs/10009754889/job/27670027039?pr=4209#step:7:37

Probably something needs an additional build tag for Windows?

**Comment by leongross** [comment]
Looks like we encountered [a similar error as a few weeks ago](https://github.com/tinygo-org/tinygo/issues/4347) @deadprogram 

```
The following packages have unmet dependencies:
 libllvm18 : Breaks: llvm-18-dev (< 1:18.1.8~++20240730104741) but 1:18.1.8~++20240730092223+3b5b5c1ec4a3-1~exp1~20240730092238.143 is to be installed
E: Unable to correct problems, you have held broken packages.
```

https://github.com/llvm/llvm-project/issues/101201

EDIT: apart from the llvm packaging failure the ci passes, would appreciate a review @ayang64 @dgryski @rminnich @deadprogram

**Comment by leongross** [comment]
llvm issues seem to be resolved by now, re-running the CI

**Comment by leongross** [comment]
CI is green, ready to merge? @aykevl @deadprogram @dgryski @rminnich

**Review by aykevl — APPROVED** [review]
Looks good to me!

**Comment by leongross** [comment]
Thanks for taking care of that @aykevl!

**Pull request merged**